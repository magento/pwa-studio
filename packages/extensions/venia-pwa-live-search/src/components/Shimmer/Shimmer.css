.card {
    width: 250px;
    margin: 10px auto;
    box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;
    padding: 20px;
}
@keyframes placeholderShimmer {
    0% {
        background-position: -468px 0;
    }

    100% {
        background-position: 468px 0;
    }
}

.shimmer-animation {
    background-color: #f6f7f8;
    background-image: linear-gradient(
        to right,
        #f6f7f8 0%,
        #edeef1 20%,
        #f6f7f8 40%,
        #f6f7f8 100%
    );
    background-repeat: no-repeat;
    background-size: 800px 104px;
    animation-duration: 1s;
    animation-fill-mode: forwards;
    animation-iteration-count: infinite;
    animation-name: placeholderShimmer;
    animation-timing-function: linear;
}

.loader {
    &-shimmer {
        &-banner {
            height: 22rem;
            background-size: 800px 22rem;
            border-radius: 5px;
            margin-bottom: 12px;
        }

        &-header {
            display: flex;
            justify-content: space-between;
            margin-bottom: 5px;
        }

        &-title {
            height: 25px;
            flex: 0 0 auto;
            width: 120px;
        }

        &-rating {
            height: 25px;
            flex: 0 0 auto;
            width: 70px;
        }

        &-list {
            height: 20px;
            width: 190px;
            margin-bottom: 5px;

            &:last-child {
                margin-bottom: 0;
            }
        }

        &-info {
            height: 20px;
            width: 220px;
            margin-bottom: 5px;

            &:last-child {
                margin-bottom: 0;
            }
        }
    }
}
