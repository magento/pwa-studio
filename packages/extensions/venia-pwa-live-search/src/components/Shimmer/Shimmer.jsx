/*
Copyright 2024 Adobe
All Rights Reserved.

NOTICE: Adobe permits you to use, modify, and distribute this file in
accordance with the terms of the Adobe license agreement accompanying
it.
*/

import React from 'react';
import { useSensor } from '../../context'; // Adjust relative path if needed

import ButtonShimmer from '../ButtonShimmer';
import FacetsShimmer from '../FacetsShimmer';
import ProductCardShimmer from '../ProductCardShimmer';

const Shimmer = () => {
  const productCardArray = Array.from({ length: 8 });
  const facetsArray = Array.from({ length: 4 });
  const { screenSize } = useSensor();
  const numberOfColumns = screenSize.columns;

  return (
    <div className="ds-widgets bg-body py-2">
      <div className="flex">
        <div className="sm:flex ds-widgets-_actions relative max-w-[21rem] w-full h-full px-2 flex-col overflow-y-auto">
          <div className="ds-widgets_actions_header flex justify-between items-center mb-md" />
          <div className="flex pb-4 w-full h-full">
            <div className="ds-sdk-filter-button-desktop">
              <button className="flex items-center bg-gray-100 ring-black ring-opacity-5 rounded-md p-sm text-sm h-[32px]">
                <ButtonShimmer />
              </button>
            </div>
          </div>
          <div className="ds-plp-facets flex flex-col">
            <form className="ds-plp-facets__list border-t border-gray-200">
              {facetsArray.map((_, index) => (
                <FacetsShimmer key={index} />
              ))}
            </form>
          </div>
        </div>
        <div className="ds-widgets_results flex flex-col items-center pt-16 w-full h-full">
          <div className="flex flex-col max-w-5xl lg:max-w-7xl ml-auto w-full h-full">
            <div className="flex justify-end mb-[1px]">
              <ButtonShimmer />
            </div>
          </div>
          <div
            className="ds-sdk-product-list__grid mt-md grid-cols-1 gap-y-8 gap-x-md sm:grid-cols-2 md:grid-cols-3 xl:gap-x-4 pl-8"
            style={{
              display: 'grid',
              gridTemplateColumns: `repeat(${numberOfColumns}, minmax(0, 1fr))`,
            }}
          >
            {productCardArray.map((_, index) => (
              <ProductCardShimmer key={index} />
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Shimmer;
