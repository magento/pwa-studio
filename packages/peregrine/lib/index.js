import * as RestApi from './RestApi';
import * as Util from './util';

// hooks
export { useEventListener } from './hooks/useEventListener';
export { useCarousel } from './hooks/useCarousel';
export { useDropdown } from './hooks/useDropdown';
export { usePagination } from './hooks/usePagination';
export { useRestApi } from './hooks/useRestApi';
export { useRestResponse } from './hooks/useRestResponse';
export { useScrollLock } from './hooks/useScrollLock';
export { useSearchParam } from './hooks/useSearchParam';
export { useSort } from './hooks/useSort';
export { useTypePolicies } from './hooks/useTypePolicies';

export {
    WindowSizeContextProvider,
    useWindowSize
} from './hooks/useWindowSize';
export { getToastId, useToasts, ToastContextProvider } from './Toasts';

// store
export { enhancer, reducers } from './store';

// components
export { default as ContainerChild } from './ContainerChild';
export { default as List, Items, Item } from './List';
export { default as Price } from './Price';
export { default as Router } from './Router';
export {
    default as PeregrineContextProvider
} from './PeregrineContextProvider';

// misc
export { RestApi };
export { Util };
export { default as createTestInstance } from './util/createTestInstance';
