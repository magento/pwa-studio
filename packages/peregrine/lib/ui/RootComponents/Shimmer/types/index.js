/**
 * This file is augmented at build time using the @magento/venia-ui build
 * target "rootShimmerTypes", which allows third-party modules to
 * add new root shimmer mappings for loading transitions.
 *
 * @see [Shimmer definition object]{@link RootShimmerTypesDefinition}
 */
export default {};

/**
 * A root shimmer definition object that describes a shimmer in your storefront.
 *
 * @typedef {Object} RootShimmerTypesDefinition
 * @property {string} shimmerType is use to map your shimmer
 * @property {string} importPath Resolvable path to the component the
 *   Root Shimmer component will render
 *
 * @example <caption>A custom root shimmer</caption>
 * const myCmsPageShimmer = {
 *      shimmerType: 'CMS_PAGE_SHIMMER',
 *      importPath: '@partner/module/path_to_your_component'
 * }
 */
