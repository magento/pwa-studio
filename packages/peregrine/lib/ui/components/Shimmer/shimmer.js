import React, { useMemo } from 'react';
import { node, number, oneOf, oneOfType, shape, string } from 'prop-types';

import { useStyle } from '../../classify';
import defaultClasses from './shimmer.module.css';

/**
 * @kind functional component
 *
 * @param {Object} classes Styles to apply to the `root` of the Shimmer. Available classes are `root` and `root_[TYPE]`.
 * @param {string|number} borderRadius Border radius for the shimmer.
 * @param {string|number} height Sets the height of the Shimmer. Numbers are in `rem` units. Strings are used directly (Example: '100px').
 * @param {string|number} width Sets the width of the Shimmer. Numbers are in `rem` units. Strings are used directly (Example: '100px').
 * @param {Object} style CSS styles to apply to the Shimmer.
 * @param {'rectangle'|'button'|'checkbox'|'radio'|'textArea'|'textInput'} type The base element shape to apply to the Shimmer.
 * @param {node} children Children to output within the Shimmer. Useful for setting image placeholders.
 */
const Shimmer = props => {
    const {
        classes: propClasses,
        borderRadius,
        height,
        width,
        style: customStyles,
        type,
        children,
        ...restProps
    } = props;
    const classes = useStyle(defaultClasses, propClasses);

    const style = useMemo(() => {
        const combinedStyles = {
            ...customStyles
        };

        Object.entries({ borderRadius, height, width }).forEach(
            ([type, value]) => {
                if (typeof value !== 'undefined') {
                    combinedStyles[type] =
                        typeof value === 'number' ? `${value}rem` : value;
                }
            }
        );

        return combinedStyles;
    }, [customStyles, borderRadius, height, width]);

    const rootClass = `root_${type}`;

    return (
        <div className={classes[rootClass]} style={style} {...restProps}>
            {children}
        </div>
    );
};

/**
 * @property {Object} classes is an object containing the class names for the
 * Shimmer component.
 * @property {string} classes.root is the class for the container
 * @property {string} classes.root_rectangle is the class for the container
 * of type rectangle
 * @property {string} classes.root_button is the class for the container
 * of type button
 * @property {string} classes.root_checkbox is the class for the container
 * of type checkbox
 * @property {string} classes.root_radio is the class for the container
 * of type radio
 * @property {string} classes.root_textArea is the class for the container
 * of type textArea
 * @property {string} classes.root_textInput is the class for the container
 * of type textInput
 * @property {number | string} borderRadius is the border radius of the Shimmer
 * @property {number | string} height is the height of the Shimmer
 * @property {number | string} width is the width of the Shimmer
 * @property {Object} style is an object of inline styles
 * @property {string} type is the type of the Shimmer
 * @property {node} children are the children of the Shimmer
 */
Shimmer.propTypes = {
    classes: shape({
        root: string,
        root_rectangle: string,
        root_button: string,
        root_checkbox: string,
        root_radio: string,
        root_textArea: string,
        root_textInput: string
    }),
    borderRadius: oneOfType([number, string]),
    height: oneOfType([number, string]),
    width: oneOfType([number, string]),
    style: shape({}),
    type: oneOf([
        'rectangle',
        'button',
        'checkbox',
        'radio',
        'textArea',
        'textInput'
    ]),
    children: node
};

Shimmer.defaultProps = {
    style: {},
    type: 'rectangle'
};

export default Shimmer;
