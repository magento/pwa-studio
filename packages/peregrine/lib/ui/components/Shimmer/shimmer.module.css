.root {
    composes: bg-gray-100 from global;
    composes: inline-block from global;
    composes: overflow-hidden from global;
    composes: pointer-events-none from global;
    composes: relative from global;
}

/* TODO @TW: cannot compose */
.root::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: var(
        --venia-global-maxWidth
    ); /* TODO @TW: review. this should use TW theme value "maxWidth.site" */
    height: 100%;
    background-image: linear-gradient(
        to right,
        rgb(var(--venia-global-color-gray-50) / 0%) 0%,
        rgb(var(--venia-global-color-gray-50)) 40%,
        rgb(var(--venia-global-color-gray-50) / 0%) 80%,
        rgb(var(--venia-global-color-gray-50) / 0%) 100%
    );
    /* composes: bg-no-repeat from global; */
    background-repeat: none;
    background-size: var(--venia-global-maxWidth) 100%;
    -webkit-animation-name: shimmerAnimation;
    -webkit-animation-duration: 1s;
    -webkit-animation-timing-function: linear;
    -webkit-animation-iteration-count: infinite;
    -webkit-animation-fill-mode: forwards;
    will-change: transform;
}

.root_rectangle {
    composes: root;

    min-height: 1rem;
}

.root_button {
    composes: root from '../Button/button.module.css';
    composes: root;

    composes: border-0 from global;
}

.root_checkbox {
    composes: input_shimmer from '../Checkbox/checkbox.module.css';
    composes: root;

    composes: border-0 from global;
}

.root_radio {
    composes: input_shimmer from '../RadioGroup/radio.module.css';
    composes: root;

    composes: border-0 from global;
}

.root_textArea {
    composes: input_shimmer from '../TextArea/textArea.module.css';
    composes: root;

    min-height: 6.75rem;
    composes: border-0 from global;
}

.root_textInput {
    composes: input_shimmer from '../TextInput/textInput.module.css';
    composes: root;

    composes: border-0 from global;
}

@-webkit-keyframes shimmerAnimation {
    0% {
        transform: translateX(-100%);
    }

    100% {
        transform: translateX(100%);
    }
}
