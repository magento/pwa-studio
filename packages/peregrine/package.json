{
  "name": "@magento/peregrine",
  "version": "12.1.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "clean": " "
  },
  "repository": "github:magento/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento/pwa-studio/tree/master/packages/peregrine#readme",
  "dependencies": {
    "fast-glob": "~3.2.4"
  },
  "devDependencies": {
    "@apollo/client": "~3.4.0",
    "@magento/eslint-config": "~1.5.0",
    "@testing-library/react-hooks": "~5.0.3",
    "intl": "~1.2.5",
    "intl-locales-supported": "~1.8.12",
    "react": "~17.0.1",
    "react-dom": "~17.0.1",
    "react-intl": "~5.20.0",
    "react-redux": "~7.2.2",
    "react-router-dom": "~5.2.0",
    "react-test-renderer": "~17.0.1",
    "redux": "~4.0.4",
    "redux-actions": "~2.6.5",
    "wait-for-expect": "~1.2.0"
  },
  "peerDependencies": {
    "@apollo/client": "~3.4.0",
    "@babel/runtime": "~7.15.3",
    "informed": "~3.29.0",
    "react": "~17.0.1",
    "react-dom": "~17.0.1",
    "react-intl": "~5.20.0",
    "react-redux": "~7.2.2",
    "react-router-dom": "~5.2.0",
    "redux": "~4.0.4",
    "redux-actions": "~2.6.5"
  },
  "engines": {
    "node": ">=14.x",
    "yarn": ">=1.12.0"
  },
  "pwa-studio": {
    "targets": {
      "declare": "./lib/targets/peregrine-declare",
      "intercept": "./lib/targets/peregrine-intercept"
    }
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
