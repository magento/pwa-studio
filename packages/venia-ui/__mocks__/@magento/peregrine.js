import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
const useTypePolicies = jest.fn(peregrine.useTypePolicies);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useEventListener,
    useDropdown,
    useSearchParam,
    useWindowSize,
    useToasts,
    useTypePolicies
};
