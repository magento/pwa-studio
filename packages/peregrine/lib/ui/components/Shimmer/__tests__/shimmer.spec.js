import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import Shimmer from '../shimmer';

jest.mock('../../../classify');
jest.mock('../../Button', () => () => <i />);

let inputProps = {};

const givenDefaultValues = () => {
    inputProps = {
        width: '100%'
    };
};

const givenWidth = () => {
    inputProps = {
        width: 200
    };
};

const givenHeight = () => {
    inputProps = {
        height: '200px'
    };
};

const givenButtonType = () => {
    inputProps = {
        type: 'button'
    };
};

const Component = () => {
    return <Shimmer {...inputProps} />;
};

describe('#Shimmer', () => {
    beforeEach(() => {
        givenDefaultValues();
    });

    it('renders default size', () => {
        const instance = createTestInstance(<Component />);

        expect(instance.toJSON()).toMatchSnapshot();
    });

    it('renders rectangle of given width', () => {
        givenWidth();
        const { root } = createTestInstance(<Component />);

        const styles = root.children[0].children[0].props.style;
        expect(styles).toHaveProperty('width', '200rem');
    });

    it('renders rectangle of given height', () => {
        givenHeight();
        const { root } = createTestInstance(<Component />);

        const styles = root.children[0].children[0].props.style;
        expect(styles).toHaveProperty('height', '200px');
    });

    it('renders Shimmer of given button type', () => {
        givenButtonType();
        const { root } = createTestInstance(<Component />);

        expect(root.children[0].children[0].props.className).toContain(
            '_button'
        );
    });
});
