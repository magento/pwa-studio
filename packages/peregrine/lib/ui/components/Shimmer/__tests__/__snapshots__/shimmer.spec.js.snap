// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Shimmer renders default size 1`] = `
<div
  className="root_rectangle"
  style={
    Object {
      "width": "100%",
    }
  }
/>
`;
